src/bson.rs

```diff
@@ -486,184 +595,243 @@ impl Bson {
-            if let (Ok(pat), Ok(opt)) = (values.get_str("$regex"), values.get_str("$options")) {
-                return Bson::Regex(Regex {
-                    pattern: pat.to_owned(),
-                    options: opt.to_owned(),
-                });
-            } else if let (Ok(code), Ok(scope)) =
-                (values.get_str("$code"), values.get_document("$scope"))
-            {
-                return Bson::JavaScriptCodeWithScope(JavaScriptCodeWithScope {
-                    code: code.to_owned(),
-                    scope: scope.to_owned(),
-                });
-            } else if let (Ok(time), Ok(increment)) = (values.get_i32("t"), values.get_i32("i")) {
-                if time >= 0 && increment >= 0 {
-                    return Bson::TimeStamp(TimeStamp {
-                        time: time as u32,
-                        increment: increment as u32,
-                    });
-                }
-            } else if let (Ok(time), Ok(increment)) = (values.get_i64("t"), values.get_i64("i")) {
-                if time >= 0
-                    && time <= std::u32::MAX as i64
-                    && increment >= 0
-                    && increment <= std::u32::MAX as i64
-                {
-                    return Bson::TimeStamp(TimeStamp {
-                        time: time as u32,
-                        increment: increment as u32,
-                    });
-                }
-            } else if let (Ok(hex), Ok(t)) = (values.get_str("$binary"), values.get_i64("type")) {
-                let ttype = t as u8;
-                return Bson::Binary(Binary {
-                    subtype: From::from(ttype),
-                    bytes: hex::decode(hex.as_bytes())
-                        .expect("$binary value is not a valid Hex encoded bytes"),
-                });
-            }
-        } else if values.len() == 1 {
-            if let Ok(code) = values.get_str("$code") {
-                return Bson::JavaScriptCode(code.to_owned());
-            } else if let Ok(hex) = values.get_str("$oid") {
-                return Bson::ObjectId(oid::ObjectId::with_string(hex).unwrap());
-            } else if let Ok(long) = values
-                .get_document("$date")
-                .and_then(|inner| inner.get_i64("$numberLong"))
-            {
-                return Bson::UtcDatetime(
-                    Utc.timestamp(long / 1000, ((long % 1000) * 1_000_000) as u32),
-                );
-            } else if let Ok(sym) = values.get_str("$symbol") {
-                return Bson::Symbol(sym.to_owned());
-            } else if let Ok(dec) = values.get_str("$numberDecimal") {
-                return Bson::Decimal128(dec.parse::<Decimal128>().unwrap());
-            } else if let Ok(undefined) = values.get_bool("$undefined") {
-                if undefined {
-                    return Bson::Undefined;
+    pub(crate) fn from_extended_document(doc: Document) -> Bson {
+        if doc.len() > 2 {
```

Comment:
I found the previous implementation of this to be rather unwieldly, so I took a stab at cleaning it up. The one downside with this new implementation is that it needs to allocate a Vec for the keys to be able to pattern match them as a slice, but this only needs to happen if there are two or fewer keys in the document, and the values are just &str, which are not very large, so I think it's fine. If we're super worried about performance here, we could also look into using [`smallvec`](https://crates.io/crates/smallvec).